fibers.cpp

```diff
@@ -238,20 +242,21 @@ bool Fibers::yield()
 		return true;
 		}
 
-	for (int i = 1; i < MAXFIBERS; ++i)
+	fi = fi % (MAXFIBERS - 1) + 1;
+	if (runnable(fibers[fi]))
 		{
-		fi = fi % (MAXFIBERS - 1) + 1;
-		if (fibers[fi].status == Fiber::ENDED)
-			{
-			// cleanup while we're searching
-			DeleteFiber(fibers[fi].fiber);
-			memset(&fibers[fi], 0, sizeof (Fiber));
-			fibers[fi].status = Fiber::REUSE;
-			}
-		else if (runnable(fibers[fi]))
+		switchto(fi);
+		return true;
+		}
+
+	for (int i = 1; i < MAXFIBERS; ++i)
+		{	
+		Fiber& f = fibers[i];
+		if (f.status == Fiber::REUSE && f.fiber)
 			{
-			switchto(fi);
-			return true;
+			LOG("delete fiber " << i << " from yield");
+			DeleteFiber(f.fiber);

```

Comment:
Although it shouldn't happen the way the code is, I think to be safe we should add an assert that we aren't deleting the current fiber.